v3/src/components/axis/hooks/use-axis.ts

```diff
@@ -17,124 +17,239 @@ export interface IUseAxis {
+        }
+        select(axisElt).selectAll('text')
+          .style('text-anchor', textAnchor)
+          .attr('transform', `${translation}${rotation}`)
+      },
 
-      // Onward to axis title
-      const
-        titleTransform = `translate(${axisBounds.left}, ${axisBounds.top})`,
-        tX = (place === 'left') ? getLabelBounds(label).height - axisGap : halfRange,
-        tY = (place === 'bottom') ? axisBounds.height - labelBounds.height / 2 : halfRange,
-        tRotation = place === 'bottom' ? '' : ` rotate(-90,${tX},${tY})`
-      if(titleRef) {
-        select(titleRef.current)
-          .selectAll('text.axis-title')
-          .data([1])
-          .join(
-            // @ts-expect-error void => Selection
-            // eslint-disable-next-line @typescript-eslint/no-empty-function
-            () => {
-            },
-            (update) => {
-              update
-                .attr("transform", titleTransform + tRotation)
-                .attr('x', tX)
-                .attr('y', tY)
-                .text(label)
-            })
+      drawAxisTitle = () => {
+        const
+          titleTransform = `translate(${axisBounds.left}, ${axisBounds.top})`,
+          tX = (place === 'left') ? getLabelBounds(label).height : halfRange,
+          tY = (place === 'bottom') ? axisBounds.height - labelBounds.height / 2 : halfRange,
+          tRotation = place === 'bottom' ? '' : ` rotate(-90,${tX},${tY})`
+        if (titleRef) {
+          select(titleRef.current)
+            .selectAll('text.axis-title')
+            .data([1])
+            .join(
+              // @ts-expect-error void => Selection
+              // eslint-disable-next-line @typescript-eslint/no-empty-function
+              () => {
+              },
+              (update) => {
+                update
+                  .attr("transform", titleTransform + tRotation)
+                  .attr('x', tX)
+                  .attr('y', tY)
+                  .text(label)
+              })
+        }
       }
 
+    // When switching from one axis type to another, e.g. a categorical axis to an
+    // empty axis, d3 will use existing ticks (in DOM) to initialize the new scale.
+    // To avoid that, we manually remove the ticks before initializing the axis.
+    select(axisElt).selectAll('.tick').remove()
+
+    scale.range(axisIsVertical ? [axisBounds.height, 0] : [0, axisBounds.width])
+    switch (type) {
+      case 'numeric':
+      case 'empty':
+        drawAxis()
+        showScatterPlotGridLines && drawScatterPlotGridLines()
+        break
+      case 'categorical':
+        drawCategoricalAxis()
+        break
     }
-  }, [axisElt, axisPlace, axis, layout, showGridLines, scale, enableAnimation,
-    place, titleRef, label, halfRange])
+    drawAxisTitle()
+  }, [axisElt, axisPlace, axis, layout, showScatterPlotGridLines, scale, enableAnimation,
+    place, titleRef, label, halfRange, type, bandWidth, collisionExists, centerCategoryLabels,])
```

Comment:
Not necessarily for this PR, but vis-a-vis our conversation about dependencies curation, virtually any change that affects the axis invalidates the `refreshAxis()` function, which in turn triggers all of the effects below, which could explain some of the extraneous axis re-renders. At some point it's worth thinking about whether the `refreshAxis()` function could be made more stable, so that the effects below could run more judiciously.

Reply:
I agree. Changes in halfRange, bandWidth and collisionExists will happen any time the graph is resized, thus triggering four of the useEffects and four calls to refreshAxis where we logically only want one.